text/0000-custom-cargo-profiles.md

```diff
@@ -0,0 +1,141 @@
+
+# Motivation
+[motivation]: #motivation
+
+Currently the "stable" way to tweak build parameters like "debug symbols", "debug assertions", and "optimization level" is to edit Cargo.toml.
+
+This file is typically checked in tree, so for many projects overriding things involves making
+temporary changes to this, which feels hacky. On top of this, if Cargo is being called by an
+encompassing build system as what happens in Firefox, these changes can seem surprising. There are
+currently two main profiles in Cargo ("dev" and "release"), and we're forced to fit everything we
+need into these two categories. This isn't really enough.
+
+Furthermore, this doesn't allow for much customization. For example, when trying to optimize for
+compilation speed by building in debug mode, build scripts will get built in debug mode as well. In
+case of complex build-time dependencies like bindgen, this can end up significantly slowing down
+compilation. It would be nice to be able to say "build in debug mode, but build build dependencies
+in release". Also, your program may have large dependencies that it doesn't use in critical paths,
+being able to ask for just these dependencies to be run in debug mode would be nice.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+Currently, the [Cargo guide has a section on this](http://doc.crates.io/manifest.html#the-profile-sections).
+
+We amend this to add that you can define custom profiles with the `profile.foo` key syntax. These can be invoked via
+`cargo build --profile foo`. The `dev`/`doc`/`bench`/etc profiles remain special. Each custom profile, aside from the
+"special" ones, gets a folder in `target/`, named after the profile. "dev" and "debug" are considered to be aliases
+
+Profile keys can be "overridden":
+
+```toml
+[profile.dev]
+opt-level = 0
+debug = true
+
+# the `image` crate will be compiled with -Copt-level=3
+[profile.dev.overrides.image]
+opt-level = 3
+
+# Dependencies semver-matching any entry in the space separated list
+# will be compiled without debuginfo
+[profile.dev.overrides."image=0.2 piston>5.0"]
+debug=false
+
+# All dependencies (but not this crate itself) will be compiled
+# with -Copt-level=2 . This includes build dependencies.
+[profile.dev.overrides."*"]
+opt-level = 2
+
+# Build scripts and their dependencies will be compiled with -Copt-level=3
+# By default, build scripts use the same rules as the rest of the profile
+[profile.dev.build_override]
+opt-level = 3
+```
+
+Custom profiles _can_ be listed in a `.cargo/config`, however the user is responsible for
+clearing up build directories if the profile changes. That is, it is undefined behavior
+to run `cargo build --profile foo` if `foo` has been defined in `.cargo/config` and the
+profile has been edited since the last time you ran `cargo build --profile foo`.
```

Comment:
How hard would it be for us to detect profile changes and automatically rebuild in that case, just as we would if you changed `Cargo.toml`?

If it's easy, can we just do it?

If it's hard, can we list it as an alternative in the RFC and explain why we're not doing it?

Reply:
It's not exactly hard, just tricky I think. We could probably do it, but I thought as a first pass it's better to leave it undefined and in the future we can experiment with a scheme for this. This introduces a dependency to files out of the project root which is kinda icky.

Reply:
We don't have to introduce a dependency on the *file*. What if we just remember the profile information used for the last compile, and if that changes, rebuild?

Reply:
Works. I can fix that.